driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -679,54 +683,49 @@ public void removeHandler(ResponseHandler handler, boolean releaseStreamId) {
-                    return;
-                }
-                handler.cancelTimeout();
-                handler.callback.onSet(Connection.this, response, System.nanoTime() - handler.startTime, handler.retryCount);
+            if (streamId < 0) {
+                factory.defaultHandler.handle(response);
+                return;
+            }
 
-                // If we happen to be closed and we're the last outstanding request, we need to terminate the connection
-                // (note: this is racy as the signaling can be called more than once, but that's not a problem)
-                if (isClosed())
-                    tryTerminate(false);
+            ResponseHandler handler = pending.remove(streamId);
+            streamIdHandler.release(streamId);
+            if (handler == null) {
+                /**
+                 * During normal operation, we should not receive responses for which we don't have a handler. There is
+                 * two cases however where this can happen:
+                 *   1) The connection has been defuncted due to some internal error and we've raced between removing the
+                 *      handler and actually closing the connection; since the original error has been logged, we're fine
+                 *      ignoring this completely.
+                 *   2) This request has timed out. In that case, we've already switched to another host (or errored out
+                 *      to the user). So log it for debugging purpose, but it's fine ignoring otherwise.
+                 */
+                streamIdHandler.unmark(streamId);
+                if (logger.isDebugEnabled())
+                    logger.debug("{} Response received on stream {} but no handler set anymore (either the request has "
+                               + "timed out or it was closed due to another error). Received message is {}", Connection.this, streamId, asDebugString(response));
+                return;
             }
+            handler.cancelTimeout();
+            handler.callback.onSet(Connection.this, response, System.nanoTime() - handler.startTime, handler.retryCount);
+
+            // If we happen to be closed and we're the last outstanding request, we need to terminate the connection
+            // (note: this is racy as the signaling can be called more than once, but that's not a problem)
+            if (isClosed())
+                tryTerminate(false);
         }
 
         @Override
-        public void channelIdle(ChannelHandlerContext ctx, IdleStateEvent e) throws Exception {
-            logger.debug("{} was inactive for {} seconds, sending heartbeat", Connection.this, factory.configuration.getPoolingOptions().getHeartbeatIntervalSeconds());
-            write(HEARTBEAT_CALLBACK);
+        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+            if (evt instanceof IdleStateEvent) {
```

Comment:
Hadn't noticed this before, but It looks like the IdleStateEvent will trigger if one side of the channel is inactive, or both sides.  We should check if it's Both sides that are idle right?  (IdleStateEvent.state() == ALL_IDLE_STATE_EVENT).    In the general case if one side is idle both will be idle due to the way our timeouts work, but would be good to check anyways.